### ISSUE TITLE
Arrow Functions incorrectly trigger errors on func-names & space-before-blocks

### ISSUE THREAD

**Issue by jesstelford** [babel/babel-eslint#27]
When I run

```bash
$ ./node_modules/.bin/eslint index.js --format=compact
```

I get the error:

```bash
index.js: line 2, col 14, Warning - Missing function expression name. (func-names)
index.js: line 2, col 14, Warning - Missing space before opening brace. (space-before-blocks)

2 problems
```

`index.js`:

```javascript
'use strict';
[1, 2, 3].map(i => i * 2);
```

## [`func-names`](https://github.com/eslint/eslint/blob/master/docs/rules/func-names.md)
The [grammar for Arrow Functions](http://wiki.ecmascript.org/doku.php?id=harmony:arrow_function_syntax) doesn't appear to allow for names. So, `func-names` should not be checked against at all.

## [`space-before-blocks`](https://github.com/eslint/eslint/blob/master/docs/rules/space-before-blocks.md)
The function is not a block (at least, according what the documentation on eslint rules considers a block).

It should be noted that this error goes away when I do actually use the block format of arrow functions:

```javascript
'use strict';
[1, 2, 3].map((i) => {
  return i * 2;
});
```

## Additional Details

`.eslintrc`:

```json
{ "parser": "babel-eslint" }
```

`package.json`:

```json
{
  ...
  "dependencies": {
    "babel": "^4.6.5",
    "babel-eslint": "babel/babel-eslint#master",
    "eslint": "^0.15.1"
  }
}
```

**Comment by sebmck** [comment]
There's only so much that can be done. Arrow functions that have expression bodies need to be transformed to have a body or else #20 happens.

**Comment by nzakas** [comment]
Can you try with the default ESLint parser to see if you get the same issues?

**Comment by keithamus** [comment]
ESLint parser does not have the same issue with func-names; [espree](https://github.com/eslint/espree) parses Arrow functions as `type: "ArrowFunctionExpression"`<sup>[[1]](https://github.com/eslint/espree/blob/4644e0561512c15e682dc27a9ac59348b7c4cb73/tests/fixtures/ecma-features/arrowFunctions/basic.src.js), [[2]](https://github.com/eslint/espree/blob/4644e0561512c15e682dc27a9ac59348b7c4cb73/tests/fixtures/ecma-features/arrowFunctions/basic.result.js)</sup> which is ignored by the func-names rule<sup>[[3]](https://github.com/eslint/eslint/blob/master/lib/rules/func-names.js#L31)</sup>

Interestingly, the Mozilla Parser API indicates that it should be neither "FunctionExpression" nor "ArrowFunctionExpression" - but actually "ArrowExpression"<sup>[[4]](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Parser_API)</sup>.

Esprima also parses arrows as "ArrowFunctionExpression"<sup>[[5]](http://esprima.org/demo/parse.html?code=()%20%3D%3E%201%3B)</sup>

**Comment by keithamus** [comment]
As for space-before-blocks. Esprima counts the inline arrow function as an expression, e.g. [`() => x*x;`](http://esprima.org/demo/parse.html?code=()%20%3D%3E%20x*x%3B) is a BinaryExpression, but does count it as a block when presented with curlies, (e.g. [`() => { return x*x; }`](http://esprima.org/demo/parse.html?code=()%20%3D%3E%20%7B%20return%20x*x%3B%20%7D)).

This behaviour seems like the correct way. Espree also follows this same pattern (I cannot link as it does not have an online repl, but I did verify this).

**Comment by sebmck** [comment]
@keithamus Yes, this is because `babel-eslint` was designed for `eslint` 0.15.x before it had ES6 support so it had to convert nodes that ESLint didn't understand to ones that it did. I'm currently tied up in work on  the Babel core but I'm more than welcome to a pull request to update all the nodes to 0.16.x.

**Comment by keithamus** [comment]
Totally fine :smile: I just thought I'd add my findings, I doubt I'll be able to find the time to fix this.

Right now I'm using eslint 0.16 with the ES6 features turned on, which seems to be more reliable than babel-eslint.

**Comment by sebmck** [comment]
Should be fixed if you upgrade `babel-eslint` to 2.0.0 and `eslint` to `0.16.x`, thanks!
